File: 20230916_Saturday/20230916.rb
# Leetcode: 1631. Path With Minimum Effort.
# https://leetcode.com/problems/path-with-minimum-effort/
# @param {Integer[][]} heights
# @return {Integer}
def minimum_effort_path(heights)
  @h = heights
  @v = Array.new(105){ Array.new(105,false)}
  @d = [[1,0],[-1,0],[0,-1],[0,1]]
  @m,@n = @h.size, @h[0].size
  l,r = 0, (10**9)+2
  while l < r
    mid = r + (r-l)/2
    @v = Array.new(105){ Array.new(105,false)}
    ok(0,0,mid)
    if @v[@m-1][@n-1]
      r = mid
    else
      l = mid + 1
    end
  end
  return l
end
def ok(i,j,mid)
  return if @v[i][j]
  @v[i][j] = true
  @d.each do |di,dj|
    ni = i + di
    nj = j + dj
    next if ! ni.between?(0,@m-1)
    next if ! nj.between?(0,@n-1)
    if (@h[i][j] - @h[ni][nj]).abs <= mid
      ok(ni,nj,mid)
    end
  end
end

File: 20230916_Saturday/20230916.py
# Leetcode: 1631. Path With Minimum Effort.
# https://leetcode.com/problems/path-with-minimum-effort/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 6073 ms, faster than 5.04% of Python3 online submissions for Path
# With Minimum Effort.
# Memory Usage: 27.3 MB, less than 5.05% of Python3 online submissions for Path
# With Minimum Effort.
# 2023.09.16 Daily Challenge.
class Solution:
    def __init__(self):
        self.vis = [[0] * 105 for _ in range(105)]
        self.dir = [(1, 0), (-1, 0), (0, -1), (0, 1)]

    def ok(self, x, y, mid, heights):
        if not self.vis[x][y]:
            self.vis[x][y] = 1
            n = len(heights)
            m = len(heights[0])

            for i in range(4):
                X = x + self.dir[i][0]
                Y = y + self.dir[i][1]

                if X < 0 or X >= n or Y < 0 or Y >= m:
                    continue

                if abs(heights[x][y] - heights[X][Y]) <= mid:
                    self.ok(X, Y, mid, heights)

    def minimumEffortPath(self, heights):
        l = 0
        r = 1000000002
        n = len(heights)
        m = len(heights[0])

        while l < r:
            mid = (l + r) // 2
            self.vis = [[0] * 105 for _ in range(105)]
            self.ok(0, 0, mid, heights)

            if self.vis[n - 1][m - 1] == 1:
                r = mid
            else:
                l = mid + 1

        return l
